### ISSUE TITLE
Pressure sensitivity steps should be finer

### ISSUE THREAD

**Issue by 4VRDriver** [xournalpp/xournalpp#1876]
**Is your feature request related to a problem? Please describe.**
The written text with pressure sensitivity turned on looks a bit unsightly because there are too few pressure steps. Especially at the end of a stroke where the pressure is almost zero, the text has weird "spikes". I uploaded an image because it's hard to describe.

**Describe the solution you'd like**
Finer pressure steps or some kind of logarithmic scale so that there are more pressure steps when the pressure is almost zero.

**Describe alternatives you've considered**
Some kind of interpolation when rendering the text. (don't know if this is even possible thought)

**Additional context**
![Screenshot_20200419_152445](https://user-images.githubusercontent.com/44267643/79689118-0a9a4400-8253-11ea-8d45-74e05e24cec4.png)
good to see on the 'z' or the bracket

**Comment by tattsan** [comment]
I don't think this is a problem of pressure steps, but of sampling.
The spikes appear when we tilt the pen quickly at the end of the strokes, because the pressure decrease rapidly.  But when we move the pen parallel while relaxing the pen, the lines get thinner smoothly at end.

**Comment by tattsan** [comment]
First, move the pen around so that you get the spikes. But if you do that more slowly, you won't get the spikes. So I think this is an issue of time interval of checking events. I don't know whether it is a problem of xournalpp or of input drivers.

**Comment by 4VRDriver** [comment]
I think you are right.

Writing text slowly in Xournalpp will make those spikes almost disappear:
Slow:
![xournal_slow](https://user-images.githubusercontent.com/44267643/88698934-fd6cdb00-d106-11ea-9ec9-6ef5c879b963.png)
Fast:
![xournal_fast](https://user-images.githubusercontent.com/44267643/88698963-09f13380-d107-11ea-9a0e-473abc3cda75.png)

In KDE Plasma there is a setting for the "raw sample rate" of the stylus. But playing around with that slider did not make a big difference at least in my case.
In comparison a (fastly) text written in Krita with same plasma-settings as in Xournalpp
![krita_fast](https://user-images.githubusercontent.com/44267643/88699457-a6b3d100-d107-11ea-91bf-981be9e15e30.png)

The text is way prettier. I don't know if Krita is checking the events faster than Xournalpp or if it just does a better job at smoothing the written strokes.
:

**Comment by tattsan** [comment]
This issue is being solved by #2729 . Let's test the PR!

**Comment by bhennion** [comment]
> This issue is being solved by #2729 . Let's test the PR!

Actually, a smoothing of pressure variations has already been merged in master (through PR #2512). See [this block of code](https://github.com/xournalpp/xournalpp/blob/master/src/control/tools/StrokeHandler.cpp#L81-L114).

This issue is probably already fixed.

**Comment by tattsan** [comment]
Thank you for your information! If the stabilizer is disabled, there will still be a slight spike. When stabilizer (arithmetic mean) is enabled, the spikes disappear completely. However, the writing taste changes considerably, so I had to reduce the buffer size.

> [But you are of course most welcome to test #2729 anyway ;-)]

I'm doing! As for the features implemented, they are perfect! :smiley:

**Comment by bhennion** [comment]
> Thank you for your information! If the stabilizer is disabled, there will still be a slight spike.

Could you make a picture, or better still, provide a .xopp file with such a stroke in it? I'll have a look.

> When stabilizer (arithmetic mean) is enabled, the spikes disappear completely. However, the writing taste changes considerably, so I had to reduce the buffer size.
> 
> > [But you are of course most welcome to test #2729 anyway ;-)]
> 
> I'm doing! As for the features implemented, they are perfect!

Thanks!

**Comment by tattsan** [comment]
Here is a comparison. Averaging method is arithmetic mean.

![image](https://user-images.githubusercontent.com/6355486/109345084-f2537680-78b2-11eb-94a5-445dca0b71fa.png)

Even if stabilizer is disabled, it is much better than before. You can see the difference in the way the thickness of the line changes with the same amount of force.
[SpikeTest.xopp.gz](https://github.com/xournalpp/xournalpp/files/6051967/SpikeTest.xopp.gz)

**Comment by bhennion** [comment]
Hmm.. I think this is due to a fairly low dpi setting. So the maximal width variation allowed is actually visible. I'm not sure this can be easily fixed.

**Comment by tattsan** [comment]
Indeed!  I changed the MAX_WIDTH_VARIATION from 0.3 to 0.06 and tried to build, then the spike went away. Is it possible to make this value vary according to the dpi value (in Xournal++'s preferences) set by the user?

**Comment by bhennion** [comment]
It sure is possible. There are possibly other constants which would benefit from being indexed on the dpi (e.g InputHandler::PIXEL_MOTION_THRESHOLD or SplineSegment::MIN_KNOT_DISTANCE/MAX_WIDTH_VARIATION).

**Comment by tattsan** [comment]
It's not a very good specification, if I do say so myself: "If the display is poor, the app's output will also be poor" .
The dpi setting should not affect the .xopp file. I'll tune up the stabilizer settings. Thank you so much.
